internal/backends/nodejs/nodejs.go

```diff
@@ -309,6 +309,84 @@ var NodejsYarnBackend = api.LanguageBackend{
 	Guess:        nodejsGuess,
 }
 
+var NodejsPNPMBackend = api.LanguageBackend{
+	Name:             "nodejs-pnpm",
+	Specfile:         "package.json",
+	Lockfile:         "pnpm-lock.yaml",
+	FilenamePatterns: nodejsPatterns,
+	Quirks: api.QuirksAddRemoveAlsoLocks |
+		api.QuirksAddRemoveAlsoInstalls |
+		api.QuirksLockAlsoInstalls,
+	GetPackageDir: func() string {
+		return "node_modules"
+	},
+	Search: nodejsSearch,
+	Info:   nodejsInfo,
+	Add: func(pkgs map[api.PkgName]api.PkgSpec, projectName string) {
+		if !util.Exists("package.json") {
+			util.RunCmd([]string{"pnpm", "init"})
+		}
+		cmd := []string{"pnpm", "add"}
+		for name, spec := range pkgs {
+			arg := string(name)
+			if spec != "" {
+				arg += "@" + string(spec)
+			}
+			cmd = append(cmd, arg)
+		}
+		util.RunCmd(cmd)
+	},
+	Remove: func(pkgs map[api.PkgName]bool) {
+		cmd := []string{"pnpm", "remove"}
+		for name := range pkgs {
+			cmd = append(cmd, string(name))
+		}
+		util.RunCmd(cmd)
+	},
+	Lock: func() {
+		util.RunCmd([]string{"pnpm", "install"})
+	},
+	Install: func() {
+		util.RunCmd([]string{"pnpm", "install"})
+	},
+	ListSpecfile: nodejsListSpecfile,
+	ListLockfile: func() map[api.PkgName]api.PkgVersion {
+		// pnpm does scoped dependencies, so only list the dependencies at the top level
+		workspaceSpecContents, err := exec.Command("pnpm", "list", "--json", "--depth", "0").Output()
```

Comment:
I imagine we actually want `--depth Inifnity` since we probably want to see all packages. This currently mirrors the specfile, I think. I'm not sure tbh what's the expected return from lock file listing, but other places list all sub dependencies.

Reply:
although this brings up a good edge condition for a future UPM improvement: UPM expects a flat list of dependencies here, but pnpm and cargo (2 off the top of my head) support hierarchical dep trees

Reply:
updated comment to reflect this state